### PR TITLE
Backpressure fix

### PR DESCRIPTION
**Pull request by lubennikovaav**

### CODE REVIEW

**Review by knizhnik — COMMENTED** [review]
There seems to be a bug in bufmgr.c:
```
if ((applyPtr != UnknownXLogRecPtr
			&& myFlushLsn > applyPtr + max_replication_apply_lag*MB))
		{
			return (myFlushLsn - flushPtr - max_replication_apply_lag*MB);
		}
```
I think applyPtr should be used in last line.

Also may be we should think about some smarter policy to throttle client: do not completly block it until gap becomes smaller than specified limit, but just slowdown. For exam;ple limit maximal wait time. Any better idea?

File: control_plane/src/compute.rs [hunk 1]
```diff
@@ -291,6 +291,9 @@ impl PostgresNode {
         conf.append("max_connections", "100");
         conf.append("wal_sender_timeout", "0");
```
Comment (ololobus) [inline control_plane/src/compute.rs:292]: In the previous PR @knizhnik set it to `10s` and wrote:

> With disabled wal_sender_timeout, walsender is not sending keep-alive messages, in response to which pageserver is able to inform about it's progress.

I haven't checked the code, though, but it seems that we have to set it to non-zero value here as well

**Comment by lubennikovaav** [comment]
> There seems to be a bug in bufmgr.c:
> 
> ```
> if ((applyPtr != UnknownXLogRecPtr
> 			&& myFlushLsn > applyPtr + max_replication_apply_lag*MB))
> 		{
> 			return (myFlushLsn - flushPtr - max_replication_apply_lag*MB);
> 		}
> ```
> 
> I think applyPtr should be used in last line.

thank you. fixed. 

> 
> Also may be we should think about some smarter policy to throttle client: do not completly block it until gap becomes smaller than specified limit, but just slowdown. For exam;ple limit maximal wait time. Any better idea?

I thought about that but didn't come up with a good heuristic and decided to keep it simple

**Comment by hlinnaka** [comment]
LockBuffer() seems like an odd place to do the sleeping. You could be holding other buffers locked already at that point, for example. How about sleeping in ProcessInterrupts()?

**Comment by lubennikovaav** [comment]
> LockBuffer() seems like an odd place to do the sleeping. You could be holding other buffers locked already at that point, for example.

Are there cases except for btree page split (and maybe other indexes) where we hold an exclusive lock on multiple pages simultaneously? Anyway, long waits for the lock may happen in a highly loaded system. I was trying to imitate this behavior.

I worry a bit about the increased number of deadlocks/backend timeouts, but this is expected if postgres is slow and means that we need to tune it better.

> How about sleeping in ProcessInterrupts()?

It will block read-only clients too, right? In my understanding, the idea of backpressure is to throttle writes and don't affect readers.

**Comment by ololobus** [comment]
Hm, just thought, should we call `CHECK_FOR_INTERRUPTS()` inside this throttle loop to properly handle interrupts during storage catch up waiting? If so, then calling `ProcessInterrupts()` inside `ProcessInterrupts()` could be not that good idea.

> Can anybody suggest some better alternative?

IIUC, any time we are going to write WAL we start critical section. So I had an idea to extend `START_CRIT_SECTION()` into something like
```clang
#define START_CRIT_SECTION() \
do { \
	if (StartCritSectionHook) { \
		StartCritSectionHook(); \
        } \
	CritSectionCount++; \
} while(0)
```

Thus, we can hook there and block only writers, but do it outside (just before) critical section at the same time. Cannot say this is a `better alternative`, but that was on my mind

**Comment by knizhnik** [comment]
+1 START_CRIT_SECTION
I also I little bit afraid to change this quite basic macro, but logically it seems to be the most relevant place:
- outside critical section
- performed on each append to WAL
- not affected by read-only transactions

But implementation should be:
{{{
#define START_CRIT_SECTION() \
do { \
	if (CritSectionCount == 0 && StartCritSectionHook) { \
		StartCritSectionHook(); \
        } \
	CritSectionCount++; \
} while(0)

**Comment by hlinnaka** [comment]
We can separate detecting that you need to sleep, from actually sleeping. We can have the check in XLogBeginInsert() where it is today, but instead of sleeping, just set a flag (a global variable in the process), and also set `InterruptPending`. Then, in ProcessInterrupts(), check if the flag is set, and sleep until the page server has caught up.

**Comment by hlinnaka** [comment]
> Hm, just thought, should we call CHECK_FOR_INTERRUPTS() inside this throttle loop to properly handle interrupts during storage catch up waiting? If so, then calling ProcessInterrupts() inside ProcessInterrupts() could be not that good idea.

That's a good point. Perhaps sleep only a little in each ProcessInterrupts() call, say 500 ms, and hope that ProcessInterrupts() gets called often enough for the page server to catch up. It's a bit unscientific, but probably enough in practice.

But yeah, in principle we should turn Processinterrupts() into a loop, so that while we're sleeping, we would also check and react to all the other conditions.

**Comment by hlinnaka** [comment]
> But yeah, in principle we should turn Processinterrupts() into a loop, so that while we're sleeping, we would also check and react to all the other conditions.

That's not very invasive if we structure it something like this:
```
diff --git a/src/backend/tcop/postgres.c b/src/backend/tcop/postgres.c
index 171f3a9500..f31a5d2db6 100644
--- a/src/backend/tcop/postgres.c
+++ b/src/backend/tcop/postgres.c
@@ -3133,6 +3133,23 @@ ProcessInterrupts(void)
 		return;
 	InterruptPending = false;
 
+	for(;;)
+	{
+		ProcesssInterrupts_pg();
+		if (backpressure_still_needed())
+		{
+			sleep_a_little();
+			continue;
+		}
+		break;
+	}
+	
+	while (backpressure_still_needed())
+}
+
+static void
+ProcesssInterrupts_pg(void)
+{
 	if (ProcDiePending)
 	{
 		ProcDiePending = false;
```

**Comment by knizhnik** [comment]
Do we actually need to block backend until back pressure condition is satisfied?
May be we should just wait for some timeout? In this case loop is not needed.
Something like this:

```
 XLogRecPtr
XLogInsert(RmgrId rmid, uint8 info)
{
      if (backpressureThrottle() > 0)
      {
           InterruptPending = true;
      }
      ...
}

void
ProcessInterrupts(void)
{
	/* OK to accept any interrupts now? */
	if (InterruptHoldoffCount != 0 || CritSectionCount != 0)
		return;
	InterruptPending = false;

        ...
        if (backpressureThrottle() > 0)
        {
            pg_usleep(backpressureTimeout);
            if (backpressureTimeout < maxBackpressureTimeout)
                backpressureTimeout *= 2;
       } else {
            backpressureTimeout = minBackpressureTimeout;
      }
}
```

**Comment by lubennikovaav** [comment]
@knizhnik , @hlinnaka , what do you think about the latest version of the PR?
There is no limit for wait time, but I think it can be done as a separate improvement later if we need it.

**Review by hlinnaka — APPROVED** [review]
I do wonder if calling `backpressure_lag` on every WAL record is too expensive. It's a pretty cheap function, but XLogInsert() is a very hot code path. Anyway, that's not new with this PR, and that's pretty easy to fix later e.g. by only checking backpressure_lag() every 100 records, so it's OK for now.

There's one "TODO: add comment" comment. And ProcessInterrupts() could use a comment to explain what it does. Other than that, looks good to me.

**Comment by knizhnik** [comment]
> @knizhnik , @hlinnaka , what do you think about the latest version of the PR? There is no limit for wait time, but I think it can be done as a separate improvement later if we need it.

Why do we reset InterruptPending in ProcessInterrupts?
Doing it in ProcessInterrupts_pg seems to be enough...

I still think that we need to provide more graceful throttling rather than block backend until pagserver catchup.
Certainly it can be done in separate PR. But if we  do not have loop, then it is not needed to introduce ProcessInterrupts_pg  functions... So I do not see much sense in doing it in two steps. Either we reject idea of waiting with increasing timeout, either we should implement it  and do not wait in loop.

**Comment by hlinnaka** [comment]
> Why do we reset InterruptPending in ProcessInterrupts? Doing it in ProcessInterrupts_pg seems to be enough...

Agreed.

On closer look:
```
		lag = backpressure_lag();
		do
		{
			ProcessInterrupts_pg();

			// Suspend writers until replicas catch up
			if (lag > 0)
			{
				elog(DEBUG2, "backpressure throttling: lag %lu", lag);
				#define BACK_PRESSURE_TIMEOUT 10000L // 0.01 sec
				pg_usleep(BACK_PRESSURE_TIMEOUT);
				lag = backpressure_lag();
			}
		} while (lag);
```
Seem strange to call backpressure_lag() before ProcessInterrupts_pg(). What if ProcessInterrupts_pg() takes a while to execute? I think this would be more natural:
```
#define BACK_PRESSURE_DELAY 10000L // 0.01 sec
		for (;;)
		{
			ProcessInterrupts_pg();

			// Suspend writers until replicas catch up
			lag = backpressure_lag();
			if (lag <= 0)
				break;
			elog(DEBUG2, "backpressure throttling: lag %lu", lag);
			pg_usleep(BACK_PRESSURE_DELAY);
		}
```

**Comment by lubennikovaav** [comment]
>Why do we reset InterruptPending in ProcessInterrupts?
Doing it in ProcessInterrupts_pg seems to be enough...

Agree. I'll fix it.

> I still think that we need to provide more graceful throttling rather than block backend until pagserver catchup. Certainly it can be done in separate PR. But if we do not have loop, then it is not needed to introduce ProcessInterrupts_pg functions... So I do not see much sense in doing it in two steps. Either we reject idea of waiting with increasing timeout, either we should implement it and do not wait in loop.

In my mental model`max_replication_flush_lag` and `max_replication_write_lag` are hard limits needed to limit the load on a very slow pageserver. This shouldn't happen during normal operation, so throttling is more like a last resort.

Besides, increasing timeout can increase rapidly (i.e. client does inserts in cycle + pageserver is totally stuck and lag only grows) and the process will sleep for too long blocking interrupts.

**Comment by knizhnik** [comment]
> In my mental modelmax_replication_flush_lag and max_replication_write_lag are hard limits needed to limit the load on a very slow pageserver. 

I afraid that it will not be true. In case of bulk load without indexes Postgres can insert data in speed more than 100Mb/sec.
I do not think that both network/safekeepers and pageserver will be able to provide such speed.
So even on normal setup bulk loading of data can require back pressure to avoid exhaustion of local disk space by written WAL.

> Besides, increasing timeout can increase rapidly (i.e. client does inserts in cycle + pageserver is totally stuck and lag only grows) and the process will sleep for too long blocking interrupts.

There are solutions for this problem:
1.Limit maximal timeout ( so increase timeout until it reaches upper bound)
2. Use WaitEvent instead of pg_sleep

**Comment by hlinnaka** [comment]
On 24/11/2021 18:51, Konstantin Knizhnik wrote:
>     In my mental modelmax_replication_flush_lag and
>     max_replication_write_lag are hard limits needed to limit the load
>     on a very slow pageserver.
> 
> I afraid that it will not be true. In case of bulk load without indexes 
> Postgres can insert data in speed more than 100Mb/sec.
> I do not think that both network/safekeepers and pageserver will be able 
> to provide such speed.
> So even on normal setup bulk loading of data can require back pressure 
> to avoid exhaustion of local disk space by written WAL.

If the safekeepers/pageserver can't keep up, IMHO we're doing something 
wrong.

>     Besides, increasing timeout can increase rapidly (i.e. client does
>     inserts in cycle + pageserver is totally stuck and lag only grows)
>     and the process will sleep for too long blocking interrupts.
> 
> There are solutions for this problem:
> 1.Limit maximal timeout ( so increase timeout until it reaches upper bound)
> 2. Use WaitEvent instead of pg_sleep

WaitEvent would indeed be nice, so that you'd see in pg_stat_activity 
what it's waiting for.

- Heikki

**Comment by knizhnik** [comment]
> If the safekeepers/pageserver can't keep up, IMHO we're doing something
wrong.
Compute node is not doing fsync-s, but safekeeper - does.
So I do not see any reasons why compute node can not insert data faster than safekeeper can persist it.

**Comment by hlinnaka** [comment]
> Compute node is not doing fsync-s, but safekeeper - does.
> So I do not see any reasons why compute node can not insert data faster than safekeeper can persist it.

That's a different thing. If the safekeepers cannot keep up with fsyncing the data, then the Postgres backend will block on commit, even without backpressure. The backpressure is for the case that the compute node *doesn't* need to wait for the commits. A bulk loading or index creation is a good example of such a workload: you don't commit until the end of the loading is finished. Or any workload with synchronous_commit=off.

**Comment by knizhnik** [comment]
> That's a different thing. If the safekeepers cannot keep up with fsyncing the data, then the Postgres backend will block on commit, even without backpressure. 

Really? Bulk load is usually done inside one transaction. So without back pressure compute node will write WAL until space on local disk is exhausted and nobody will stop it.

**Comment by knizhnik** [comment]
Sorry, I replied faster than read your message till the end.
But my point is that situation when compute node produce WAL faster than safekeeper/pageserver can consume it is not something abnormal or exotic. It should happen quite frequently: during bulk load, large transations, large number of concurrent transactions... In all cases when backends can together generate a lot of WAL before committing their transactions.

**Comment by hlinnaka** [comment]
On 25/11/2021 11:24, Konstantin Knizhnik wrote:
> Sorry, I replied faster than read your message till the end.
> But my point is that situation when compute node produce WAL faster than 
> safekeeper/pageserver can consume it is not something abnormal or 
> exotic. It should happen quite frequently: during bulk load, large 
> transations, large number of concurrent transactions... In all cases 
> when backends can together generate a lot of WAL before committing their 
> transactions.

Maybe we're meaning different things by normal and abnormal. The 
safekeeper and pageserver run on big servers and they serve multiple 
tenants. A single tenant should not be able to overload them, no matter 
what it does, no matter what the workload is. That's what I mean by 
normal circumstances.

That said, things happen. Maybe we didn't tune things right. Maybe 
there's a bug in our code. Maybe a disk is failing. Maybe the tenant had 
been inactive for a month and we had "swapped it out" to S3 and it takes 
a while to load stuff back into the page server. Maybe a pageserver 
crashed and we just started a new one, and it's very slow before it gets 
fully warmed up. Maybe there's a spike where 100 different tenants 
become active at the same time. Things like that. We need the 
backpressure to deal with that kind of stuff.

- Heikki

**Comment by lubennikovaav** [comment]
> > In my mental modelmax_replication_flush_lag and max_replication_write_lag are hard limits needed to limit the load on a very slow pageserver.
> 
> I afraid that it will not be true. In case of bulk load without indexes Postgres can insert data in speed more than 100Mb/sec. I do not think that both network/safekeepers and pageserver will be able to provide such speed. So even on normal setup bulk loading of data can require back pressure to avoid exhaustion of local disk space by written WAL.

Thank you for the clarification. So we have two totally different use cases for backpressure:
1. Normal zenith operation + too fast bulk inserts.
2. Stuck pageserver (or wal service) + normal write load on compute node.

In any case, for the process to converge, we need to hold writes until lag decreases.

What about something like this?

```
#define BACK_PRESSURE_DELAY 10000L // 0.01 sec

prev_lag = 0;
for (;;)
    {
        ProcessInterrupts_pg();

        lag = backpressure_lag();

        // Replicas totally caught up. We're done here.
        if (lag <= 0)
            break;

        // Replicas caught up a bit. This round of throttling is done.
        // If replicas still lag, we'll do more throttling in the next interrupt handler.
        if (prev_lag - lag > 1MB) // TODO: How to chose this distance?
            break;

        elog(DEBUG2, "backpressure throttling: lag %lu", lag);
        pg_usleep(BACK_PRESSURE_DELAY);
        prev_lag = lag;
    }
```

**Comment by knizhnik** [comment]
> A single tenant should not be able to overload them

So what people are doing when create new database?
Them are populating it with data. Doing something like pgbench -i
It is definitely normal operation.
My point is that without back-pressure this normal operation can cause generation of too much WAL at compute node.
It happens not because "shit happens" or there is something wrong with our configuration or S3 or bug in out code.
It is just normal and expected behavior, because compute node is not performing fsyncs, but safekeer has to perform them.
At my notebook the difference of pgbench -i -s 100 with fsync=on/off is about 2 times.

I am not sure how critical this behavior is for Zenith. Taken in account that maximal size of table is also limited now by size of local storage. I just think that we should consider back-pressure as some kind of "last resort",  but mechanism which quite frequently will be involved. So its work should be delicate and not causing collapse of client applications errors and restarts because of timeout expirationar .

**Comment by hlinnaka** [comment]
On 25/11/2021 12:21, Konstantin Knizhnik wrote:
>     A single tenant should not be able to overload them
> 
> So what people are doing when create new database?
> Them are populating it with data. Doing something like pgbench -i
> It is definitely normal operation.

Right.

> My point is that without back-pressure this normal operation can cause 
> generation of too much WAL at compute node.

I disagree with that.

It might well happen today, I wouldn't be surprised, but if so that is 
an issue we should fix.

> It happens not because "shit happens" or there is something wrong with 
> our configuration or S3 or bug in out code.
> It is just normal and expected behavior, because compute node is not 
> performing fsyncs, but safekeer has to perform them.
> At my notebook the difference of pgbench -i -s 100 with fsync=on/off is 
> about 2 times.

That's because the COMMIT is waiting for the safekeeper to flush the 
data, right? Backpressure doesn't kick in in that scenario.

/me tests that with this PR

You're right, it *does* kick in a lot. Hmm, why is that?

First observation: in 'main', the pageserver performs layer eviction in 
the WAL receiver thread. That causes hiccups that last about 0.5 seconds 
on my laptop. That's enough for the backpressure to kick in.

So I switch to testing this with 
https://github.com/zenithdb/zenith/pull/864. With that, the backpressure 
*still* kicks in regularly. Why is that?

The default max_replication_flush_lag is 16MB. I think that is simply 
too small. That's exactly the size of a Postgres WAL segment. If you are 
bulk loading data and there are no commits, Postgres will only flush the 
WAL when switching a WAL segment, that is, every 16 MB. If the max lag 
is also 16 MB, you're pretty much guaranteed to stop every time.

I bumped max_replication_flush_lag to 128 MB, and that mostly stopped 
the "flush lag" from being triggered. It still happens in the "create 
index" phase a little, not sure why. Maybe even 128 MB is just too 
small, or maybe the pageserver checkpoint causes a big enough hiccup on 
my laptop to affect the WAL receiver. We should investigate that.

However, "apply lag" is still triggered. The default for that is 1 GB, 
and I think that's also too small, for the same reason that 16 MB is too 
small for the flush lag. 1 GB is exactly the default 
checkpoint_distance, so you are again pretty much guaranteed to hit that 
limit. I bumped up "max_replication_apply_lag" to "1500MB", and now the 
"apply lag" isn't triggered anymore.

This testing was without safekepeers. With safekeepers, the backpressure 
still kicks in a lot. The WAL isn't reaching the page server fast 
enough. I don't know why but that clearly needs to be fixed, regardless 
of backpressure. Without backpressure, you get timeouts at 
get_page_at_lsn() requests instead, so the backpressure seems to be 
working as intended to prevent that from happening.

> I am not sure how critical this behavior is for Zenith. Taken in account 
> that maximal size of table is also limited now by size of local storage. 
> I just think that we should consider back-pressure as some kind of "last 
> resort", but mechanism which quite frequently will be involved. So its 
> work should be delicate and not causing collapse of client applications 
> errors and restarts because of timeout expirationar .

Agreed, smoother behavior is always nicer, even if we don't expect it to 
happen under normal circumstances. We certainly don't want application 
errors or timeouts. Do you think this PR would lead to errors as it is?

- Heikki

**Comment by knizhnik** [comment]
> > > In my mental modelmax_replication_flush_lag and max_replication_write_lag are hard limits needed to limit the load on a very slow pageserver.
> > 
> > 
> > I afraid that it will not be true. In case of bulk load without indexes Postgres can insert data in speed more than 100Mb/sec. I do not think that both network/safekeepers and pageserver will be able to provide such speed. So even on normal setup bulk loading of data can require back pressure to avoid exhaustion of local disk space by written WAL.
> 
> Thank you for the clarification. So we have two totally different use cases for backpressure:
> 
>     1. Normal zenith operation + too fast bulk inserts.
> 
>     2. Stuck pageserver (or wal service) + normal write load on compute node.
> 
> 
> In any case, for the process to converge, we need to hold writes until lag decreases.
> 
> What about something like this?
> 
> ```
> #define BACK_PRESSURE_DELAY 10000L // 0.01 sec
> 
> prev_lag = 0;
> for (;;)
>     {
>         ProcessInterrupts_pg();
> 
>         lag = backpressure_lag();
> 
>         // Replicas totally caught up. We're done here.
>         if (lag <= 0)
>             break;
> 
>         // Replicas caught up a bit. This round of throttling is done.
>         // If replicas still lag, we'll do more throttling in the next interrupt handler.
>         if (prev_lag - lag > 1MB) // TODO: How to chose this distance?
>             break;
> 
>         elog(DEBUG2, "backpressure throttling: lag %lu", lag);
>         pg_usleep(BACK_PRESSURE_DELAY);
>         prev_lag = lag;
>     }
> ```

So as far as I understand you want to distinguish sitations when zenith storage (safekeeper, pagerserver, S3,...) are slower than compute node and when zenith is stuck/unavailable. In first case we will get some positive feedback (gap is decreased), in the second - no. I am not sure if it is really needed to distinguish this two situations and handle them in different way.

First of all I wan to notice that compute node (walproposer) receives feed back from replica with some delay. If there are no write transactions, then walsender sends keep-aive message after expiration of wal_sender_timeout / 2.
So if it is now set to 10 seconds, then next update of status (when we get updated information about lag) will happen not earlier than after 5 seconds.  So your proposal to exit loop on positive dynamic will not work in this case and backend still be blocked for 5 seconds which seems to be to long for most application.

Second, as far as InsertRecord is called very frequently. So most likely we detect that lags exceeds threshold very fast, i.e. lets threshold be 10Gb. When gap becomes 1Gb + 100 bytes we detect this situation and start sleeping. When we are awoken, if pageserver really do some progress, them lag will be less than threshold. In other words: if pageserver+safekeeeper are alive and doing some useful work, consuming WAL, then after sleep lag will be less then threshold. So there is no need to check for positive dynamic: if there is positive dynamic than we are beyond threshold and extra check is not needed.

Third, assume that we have min_bp_timeout...max_bp_timeout range for back-pressure timeout. Let's max_bp_timeout == 1 second.
So it meas that we can not insert more than 1 WAL record in second. Even if safekeeper/pagerserver are stuck, there is no chance to cause local WAL overflow with such rate of writes (taken in account that average WLA record size is less than 100 bytes and even with FPI, it is still less than 1Gb per day. Even with limited local storage it seems to be not noticeable amount.

So I still think that exponentially increased timeout with limited upper bound and no loop around it will be better solution. And pg_usleep should be replaced with WaitEvent

**Comment by lubennikovaav** [comment]
> So as far as I understand you want to distinguish sitations when zenith storage (safekeeper, pagerserver, S3,...) are slower than compute node and when zenith is stuck/unavailable. In first case we will get some positive feedback (gap is decreased), in the second - no. I am not sure if it is really needed to distinguish this two situations and handle them in different way.
> 

I suggest handling them in the same way. 

> First of all I wan to notice that compute node (walproposer) receives feed back from replica with some delay. If there are no write transactions, then walsender sends keep-aive message after expiration of wal_sender_timeout / 2. So if it is now set to 10 seconds, then next update of status (when we get updated information about lag) will happen not earlier than after 5 seconds. So your proposal to exit loop on positive dynamic will not work in this case and backend still be blocked for 5 seconds which seems to be to long for most application.

Hmm, good point.
So there is no sense in rechecking backpressure_lag() more often than `wal_sender_timeout / 2`, right?
Should we just tune feedback to be more frequent than 5s?

> Second, as far as InsertRecord is called very frequently. So most likely we detect that lags exceeds threshold very fast, i.e. lets threshold be 10Gb. When gap becomes 1Gb + 100 bytes we detect this situation and start sleeping. 
When we are awoken, if pageserver really do some progress, them lag will be less than threshold. In other words: if pageserver+safekeeeper are alive and doing some useful work, consuming WAL, then after sleep lag will be less then threshold. So there is no need to check for positive dynamic: if there is positive dynamic than we are beyond threshold and extra check is not needed.

Note that now we sleep in a very small portions, to be able to handle interrupts. So most likely, there is no progress in between two periods of sleep. With WaitEvent implementation, it won't be necessary, though.

> Third, assume that we have min_bp_timeout...max_bp_timeout range for back-pressure timeout. Let's max_bp_timeout == 1 second. So it meas that we can not insert more than 1 WAL record in second. Even if safekeeper/pagerserver are stuck, there is no chance to cause local WAL overflow with such rate of writes (taken in account that average WLA record size is less than 100 bytes and even with FPI, it is still less than 1Gb per day. Even with limited local storage it seems to be not noticeable amount.

Does it mean that the`current_bp_timeout` is global?
Otherwise, retrying client that always starts from `min_bp_timeout` will be able to generate a lot of data

**Comment by knizhnik** [comment]
> That's because the COMMIT is waiting for the safekeeper to flush the
data, right? Backpressure doesn't kick in in that scenario.

No!
As I already wrote everything happens before commit.
Please notice, that I am talking now not about Zenith, but just about vanilla Postgres.
I do pgbench -i -s 100 and tuples are inserted twice faster with fsync=off then with fsync=on (87 sec vs. 187sec).
Why? Frankly speaking I do not know. Yes, postgres is performing some fsyncs before commit, just because it has to flush WAL if some page is evicted. But according to profiler, we spent in fsync only 1.9% of time.

But it is a fact: with fsync=off we are twice faster and insert 16Gb of data in 87 seconds (188Mb/sec).
Will safekeeper be be to provide the same speed if it has o perform fsync after each 128kb?
At my notebook just writing 128kb block followed by fsync shows speed 26Mb/sec. So even without safekeeper and network overhead, compute node is producing data 6 times faster than safekeeper is able to consume!

**Comment by hlinnaka** [comment]
On 25/11/2021 15:41, Konstantin Knizhnik wrote:
>     That's because the COMMIT is waiting for the safekeeper to flush the
>     data, right? Backpressure doesn't kick in in that scenario.
> 
> No!
> As I already wrote everything happens before commit.
> Please notice, that I am talking now not about Zenith, but just about 
> vanilla Postgres.
> I do pgbench -i -s 100 and tuples are inserted twice faster with 
> fsync=off then with fsync=on (87 sec vs. 187sec).
> Why? Frankly speaking I do not know. Yes, postgres is performing some 
> fsyncs before commit, just because it has to flush WAL if some page is 
> evicted.

Hmm, wal_buffers fill up, so it needs to wait for the XLogWrite() to 
finish before more WAL can be inserted?

Was this with shared_buffers='1 MB'? That auto-tunes wal_buffers to just 
64 kB if it's not specified xplicitly, which could easily become a 
bottleneck. We should bump that up..

> But according to profiler, we spent in fsync only 1.9% of time.

Are you sure that's counting wall-clock time that includes time waiting 
for I/O or locks, and not merely active CPU time? Plain 'perf record' 
only records active CPU time.

> But it is a fact: with fsync=off we are twice faster and insert 16Gb of 
> data in 87 seconds (188Mb/sec).
> Will safekeeper be be to provide the same speed if it has o perform 
> fsync after each 128kb?
> At my notebook just writing 128kb block followed by fsync shows speed 
> 26Mb/sec. So even without safekeeper and network overhead, compute node 
> is producing data 6 times faster than safekeeper is able to consume!

pgbench initialization should definitely not wait for fsync every 128 kB 
of WAL. Is that because of the small wal_buffers? Although that 
shouldn't force an fsync either, just writing it out. Hmm, unless you're 
using wal_sync_methd='open_sync' or 'open_datasync'. 'fdatasync' is the 
default on my Linux laptop, but maybe it's different on your system?

- Heikki

**Comment by knizhnik** [comment]
> Do you think this PR would lead to errors as it is?
So if we block backend for long enough  time (more than once second), then most clients will interrupt request due too timeout expiration. And either try to reestablish new connection with server and repeat query (just incresing system workload), 
either report error to user.

The question is whether we we can block client for so long time when all components of Zenith are working normally?
During bulk load we should detect lagging very fast, it means that lag should not become significantly larger than throttle threshold. And it should not take too much time to make this lag smaller then threshold. So backend should not be blocked for a long time. But only if CHECK_FOR_INTERRUPTS is called frequent enough...

But if something happen with Zenith components, then infinite waiting until lag not exceeds threshold (or at least that lag is decreased) will cause failure of application. But if we limit maximal delay, then ... the question is how much WAL backend can produce before CHECK_FOR_INTERRUPTS. If it is not so large, then we can avoid WAL overflow and still provide reasonable response time at least for short transactions.

Sorry for so long discussion which I have initiated:(
May be can really implement in this PR the simplest solution and leave more investions and debates for a  future.

**Comment by knizhnik** [comment]
> Hmm, wal_buffers fill up, so it needs to wait for the XLogWrite() to
finish before more WAL can be inserted?

When page is evicted from buffer cache, then we should guarantee than WAL is flushed until it's LSN.
So this is why we Postgres periodically calls fsync:
```
#0  XLogFlush (record=10272500728) at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/transam/xlog.c:2906
#1  0x000055640551741c in FlushBuffer (buf=buf@entry=0x7fcae69e2940, reln=0x556405ee0a18, reln@entry=0x0)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/storage/buffer/bufmgr.c:2922
#2  0x00005564055187aa in FlushBuffer (reln=0x0, buf=0x7fcae69e2940) at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/storage/buffer/bufmgr.c:2873
#3  BufferAlloc (foundPtr=0x7fffe724b82b, strategy=0x556405f0f478, blockNum=896854, forkNum=MAIN_FORKNUM, relpersistence=112 'p', smgr=0x556405ee0a18)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/storage/buffer/bufmgr.c:1265
#4  ReadBuffer_common (smgr=0x556405ee0a18, relpersistence=<optimized out>, forkNum=MAIN_FORKNUM, blockNum=896854, mode=RBM_ZERO_AND_LOCK, strategy=0x556405f0f478, 
    hit=0x7fffe724b8d7) at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/storage/buffer/bufmgr.c:863
#5  0x0000556405518d8a in ReadBufferExtended (reln=reln@entry=0x7fcae647cc40, forkNum=forkNum@entry=MAIN_FORKNUM, blockNum=blockNum@entry=4294967295, 
    mode=mode@entry=RBM_ZERO_AND_LOCK, strategy=<optimized out>) at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/storage/buffer/bufmgr.c:770
#6  0x00005564052a6e9e in ReadBufferBI (relation=relation@entry=0x7fcae647cc40, targetBlock=targetBlock@entry=4294967295, mode=mode@entry=RBM_ZERO_AND_LOCK, 
    bistate=bistate@entry=0x556405e705a0) at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/heap/hio.c:120
#7  0x00005564052a742a in RelationGetBufferForTuple (relation=relation@entry=0x7fcae647cc40, len=128, otherBuffer=otherBuffer@entry=0, options=options@entry=2, 
    bistate=bistate@entry=0x556405e705a0, vmbuffer=vmbuffer@entry=0x7fffe724ba6c, vmbuffer_other=0x0)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/heap/hio.c:624
#8  0x000055640529b069 in heap_multi_insert (relation=<optimized out>, slots=0x556405f25a18, ntuples=1000, cid=<optimized out>, options=2, bistate=0x556405e705a0)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/heap/heapam.c:2381
#9  0x00005564053730ed in table_multi_insert (bistate=<optimized out>, options=<optimized out>, cid=<optimized out>, nslots=1000, slots=0x556405f25a18, rel=<optimized out>)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/include/access/tableam.h:1429
```

**Comment by knizhnik** [comment]
> Was this with shared_buffers='1 MB'? That auto-tunes wal_buffers to just
64 kB if it's not specified xplicitly, which could easily become a
bottleneck. We should bump that up..

Yes, this is 1Mb shared buffers. With such small shared buffers I tried to emulate behavior of Postres with much larger then few Gb data sets. Certainly, as as any simulation, it can be not so precise.

> Plain 'perf record' only records active CPU time.
I am build perf --call-graph dwarf profile. It shows percent of total spent time, not just CPU time (at least if I write program which in the loop calls two function and first sleeps for 1 second and another - for two seconds, then profiles shows that I spent twice more time in the second function).

> pgbench initialization should definitely not wait for fsync every 128 kB of WAL.
128kb is maximal size of messag sent by walsender. So in ideal case (bulk load or large number of transactions) safekeeper is receiving data in 128kb chunks and does fsync after writing this chunk to the disk and before sending reply to the walporposer.
 It can be optimized but ... this is the current behavior of safekeeper.

**Comment by lubennikovaav** [comment]
One more update to the PR. 
 @hlinnaka , @knizhnik , what do you think about merging this version and tuning it later in some subsequent PRs?

**Comment by hlinnaka** [comment]
> One more update to the PR. @hlinnaka , @knizhnik , what do you think about merging this version and tuning it later in some subsequent PRs?

+1

**Comment by lubennikovaav** [comment]
While on it, I finally understood, what bothers me about the backpressure limits - I don't understand how we tune the values.

We have a timeout in wait_lsn() which throws the error if wal streaming is too slow (bottleneck can be either in wal service or in pageserver's walreceiver):
` Timed out while waiting for WAL record at LSN 0/9396FE0 to arrive, last_record_lsn=0/8697FB8, disk consistent LSN=0/169ADB0: SeqWaitError`

I observe it from time to time in `performance/test_bulk_insert.py` on my laptop.

In my understanding, backpressure should save us from this error, by throttling the write load to some digestible amount.
But if we set the `lag` too large it will be already too late for throttling, when we hit it.

What do you think?

**Comment by knizhnik** [comment]
There are two reasons for using back pressures:
1. Prevent local disk overflow with generated WAL
2. Limit response time of get_page_at_lsn

First limit we can specify either statically (looks at the ceiling and choose some value  which we assume reasonable for WAL per  tenant), either dynamically: divide available disk space by number of tenant.
Second threshold depends  speed of WAL replaying. Just find out how much WAL we can replay let's say in one second. Applying UPDATE/INSERT WAL records takes about 2usec and its size is about 100 bytes. It leads to 50Mb WAL pressure threshold. It seems to be small enough so that we can completely ignore 1)

Certainly speed of replaying WAL depends on many factors. But I hope we can choose proper thresholds values using our monitoring tool (particularly looking at page_at_lsn max/average time)

**Comment by hlinnaka** [comment]
> 



> 



> > Hmm, wal_buffers fill up, so it needs to wait for the XLogWrite() to
> > finish before more WAL can be inserted?
> 
> When page is evicted from buffer cache, then we should guarantee than WAL is flushed until it's LSN. So this is why we Postgres periodically calls fsync:

Good point. I think we could skip that in Zenith. AFAICS, there's no particular need to wait for the WAL flush before evicting in Zenith.

**Pull request merged**